### PR TITLE
Feature decouple session data layer

### PR DESCRIPTION
**Pull request by sthenusan**
### Proposed changes in this pull request

Currently, the identity server user session implementation is backed by relational databases. That implementation will affect the performance in large scale implementations. This pull request contains POC to the session data is backed by Redis server. Here Both Redis and RDMS support is given.

#### Code

The persistSessionData, getSessionContextData methods in the Session store class are modified for the change. redisEnabled variable is used to select either the Redis implementation or RDMS implementation.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 1]
```diff
@@ -488,15 +480,81 @@ public void removeExpiredOperationData() {
...
+            byte[] sessionObject = jedis.get(key.getBytes());
+            if (sessionObject != null) {
+                HashMap hash = (HashMap) deserialize(sessionObject);
+                String operation = (String) hash.get(OPERATION);
+                long nanoTime = ((Long) hash.get(NANO_TIME));
+
+                if ((OPERATION_STORE.equals(operation))) {
+                    Object blobObject = hash.get(BLOB_OBJ);
+                    return new SessionContextDO(key, type, blobObject, nanoTime);
+
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:500]: Excess line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 2]
```diff
@@ -191,7 +198,21 @@
...
+    private final String sqlInsertDELETE;
+    private String sqlDeleteSTORETask;
+    private final String sqlDeleteTempDataTask;
+    private final String sqlDeleteDELETETask;
+    private String sqlSelect;
+    private String sqlDeleteExpiredDataTask;
+    private int deleteChunkSize = DEFAULT_DELETE_LIMIT;
+    private boolean sessionDataCleanupEnabled = true;
+    private boolean operationDataCleanupEnabled = false;
+    private final boolean redisEnabled = true;
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:212]: Where are we reading this from the property file?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 3]
```diff
@@ -191,12 +198,33 @@
...
+    private final String sqlInsertDELETE;
+    private String sqlDeleteSTORETask;
+    private final String sqlDeleteTempDataTask;
+    private final String sqlDeleteDELETETask;
+    private String sqlSelect;
+    private String sqlDeleteExpiredDataTask;
+    private int deleteChunkSize = DEFAULT_DELETE_LIMIT;
+    private boolean sessionDataCleanupEnabled = true;
+    private boolean operationDataCleanupEnabled = false;
+    private boolean redisEnabled = true;
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:212]: Let's set the default value to be false.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 4]
```diff
@@ -191,12 +198,33 @@
...
     private SessionDataStore() {
+
         String enablePersistVal = IdentityUtil.getProperty("JDBCPersistenceManager.SessionDataPersist.Enable");
         enablePersist = true;
         if (enablePersistVal != null) {
             enablePersist = Boolean.parseBoolean(enablePersistVal);
         }
+
+        String redisEnabledVal = IdentityUtil.getProperty("JDBCPersistenceManager.SessionDataPersist.RedisEnable");
+        if (redisEnabledVal != null) {
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:223]: Let's use StringUtils.isNotBlank() here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 5]
```diff
@@ -191,12 +198,33 @@
...
+            log.info("checking");
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:224]: Let's add a comprehensive log message here.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 6]
```diff
@@ -288,70 +317,39 @@ public static SessionDataStore getInstance() {
         return instance;
     }
 
+    // Return Jedis object.
+    private static BinaryJedis getJedisInstance() {
+
+        if (jedis != null)
+            return jedis;
+        return new BinaryJedis(HOST, PORT);
+    }
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:326]: Can we guarantee that this new instance creation would be thread-safe?  Shall we extract this out and use the singleton pattern here? WDYT?

Comment (sthenusan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:326]: yeah aiya. Using singleton is better here.

**Review by ChamathNS — APPROVED** [review]

**Pull request merged**